compiler_opt/tools/regalloc_trace/extract_functions_lib.py

```diff
@@ -0,0 +1,163 @@
+    function_name: The (mangled) name of the function to extract.
+    llvm_extract_path: The path to the llvm-extract binary to use to extract
+      the function.
+    opt_path: The path to the opt binary to use to strip debug information.
+  """
+  command_vector = [
+      llvm_extract_path,
+      "-func",
+      function_name,
+      "-o",
+      output_file_path + ".fat",
+      bc_file_path,
+  ]
+  subprocess.run(command_vector, capture_output=True, check=True)
+
+  opt_command_vector = [
+      opt_path,
+      "-strip-debug",
+      "-o",
+      output_file_path,
+      output_file_path + ".fat",
+  ]
+  subprocess.run(opt_command_vector, capture_output=True, check=True)
+  os.remove(output_file_path + ".fat")
+
+  orig_cmd_file_path = bc_file_path[:-3] + ".cmd"
+  output_cmd_file_path = output_file_path[:-3] + ".cmd"
+  shutil.copy(orig_cmd_file_path, output_cmd_file_path)
+
+
+def get_function_module_map(corpus_path: str,
+                            llvm_nm_path: str) -> dict[str, str]:
+  """Gets a mapping from function names to module paths.
+
+  Args:
+    corpus_path: The path to the corpus to obtain the mapping from.
+    llvm_nm_path: The path to the llvm-nm binary to obtain symbols from
+      bitcode files.
+
+  Returns:
+    A dictionary mapping (mangled) function names to module paths.
+  """
+  function_to_module_map = {}
+
+  with open(
+      os.path.join(corpus_path, "corpus_description.json"),
+      encoding="utf-8") as corpus_description_handle:
+    corpus_description = json.load(corpus_description_handle)
+
+  for module in corpus_description["modules"]:
+    module_path = os.path.join(corpus_path, module) + ".bc"
+    for function_name in _get_functions_in_file(module_path, llvm_nm_path):
+      function_to_module_map[function_name] = module_path
+
+  return function_to_module_map
+
+
+def extract_functions(functions_to_extract: Sequence[str],
+                      function_to_module: dict[str, str],
+                      llvm_extract_path: str, opt_path: str, thread_count: int,
+                      output_dir: str) -> None:
+  """Extracts all the functions specified.
+
+  Args:
+    functions_to_extract: A string list containing (mangled) names of all the
+      functions that should be extracted.
+    function_to_module: A dictionary mapping (mangled) function names to module
+      paths.
+    llvm_extract_path: The path to the llvm-extract binary to use to extract
+      function bodies from bitcode files.
+    opt_path: The path to the opt binary to use for stripping debug info.
+    thread_count: The number of threads to use for extracting functions.
+    output_dir: The path to the new corpus where all the extracted functions
+      will be placed.
+  """
+  module_paths = []
+
+  with concurrent.futures.ThreadPoolExecutor(thread_count) as thread_pool:
+    extract_futures = []
+
+    for index, function_to_extract in enumerate(functions_to_extract):
```

Comment:
(nit / question) instead of `index`, why not output to a file with the module and function name in the name? really large function names?

is the numbering scheme deterministic (for debugging) - like is it in order of occurrence in module? may be worth logging or something the index -> function association, per module

Reply:
It keeps the file names easy to deal with, there isn't really any point in denoting what functions are in what files with how the rest of the infra works (and it makes even less sense when we start grouping functions), and it isn't hard to figure it out for debugging purposes.

The numbering is deterministic.